### PR TITLE
Run 10 Phase 1: publish BEFORE fallback gate

### PR DESCRIPTION
**Pull request by EtanHey**
## Summary

- publish the accepted Run 10 Phase 1 BEFORE receipt without remeasurement
- make every observed MCP-side CLI fallback fail the benchmark process and emit `ci_gate` receipt evidence
- render compact non-PASS/moved rows, a collapsed full table, measurement tags, explicit gate reasons, D180 transport trust, and conditional D201 mechanism evidence

## Scope

- no `src/server.ts`, `src/agent-engine.ts`, socket-routing, or paste-path changes
- accepted receipt remains immutable
- no new P0/P1 found versus main

## Verification

- `bun test tests/bench-e2e.test.ts`: 113 passed
- `bun run test`: 155 files, 3,696 passed, 1 skipped, exit 0
- unset `CMUX_SOCKET_PATH` / `CMUX_DAEMON_SOCKET` full suite: same 155 / 3,696 / 1, exit 0
- `bun run typecheck` in both environment modes: exit 0
- D138 `tests/no-machine-home-literals.test.ts`: 1 passed
- protected push hook: 155 files, 3,696 passed, 1 skipped, exit 0
- local CodeRabbit repair review: zero findings
- Claude pair-review: actionable findings repaired and re-reviewed

CI note: GitHub CI exercises the fallback classifier through Vitest. The live benchmark itself requires isolated cmux and is not run by GitHub Actions; when invoked, the benchmark process exits nonzero on any MCP CLI fallback and records the reasons in `ci_gate`.

— cmuxlayerCodex-2c968dd2 (worker) · codex/gpt-5.6-sol

---

> [!NOTE]
> **Medium Risk**
> Behavior change: benchmark runs that previously exited 0 can now fail on MCP transport gating; impact is limited to the bench harness and publication tooling, not production MCP routing.
> 
> **Overview**
> Adds a **benchmark CI gate** in `scripts/bench-e2e.mjs`: exported `benchmarkGateFailures` flags fatal errors, row operation errors, MCP **CLI fallback** evidence (inferred transport, transport counts, fallback counts), and **unattested transport** on MCP rows, with an exception for D180 untrusted `send_to` surface rows. The harness **exit code** now follows any gate failure instead of only `fatal_error` / `error_count`, and JSON receipts gain **`ci_gate`** (`verdict` + `failures`).
> 
> Also adds **`renderPhase1BeforePublication`** to turn an immutable receipt into Run 10 Phase 1 BEFORE PR-comment markdown: non-PASS summary table, collapsed full matrix, `sampled` / `single_shot` tags, gate reason codes, D180 **UNTRUSTED** transport labeling, and conditional **D201** mechanism evidence (or INCONCLUSIVE when predicates or rows are incomplete).
> 
> **Vitest** in `tests/bench-e2e.test.ts` locks in five fallback/unattested scenarios and publication/D201 edge cases. A **implementation plan** doc is added under `docs/plans/`; no server or agent-engine changes.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 1d758552686944a86fa2d749187b220096add0a1. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

> [!NOTE]
> ### Add benchmark gate and Phase 1 BEFORE publication renderer to `scripts/bench-e2e.mjs`
> - Adds `benchmarkGateFailures` which classifies benchmark rows as failures for fatal errors, per-row operation errors, MCP CLI fallback evidence, and unattested transports
> - Adds `renderPhase1BeforePublication` which produces a markdown publication artifact with visible/collapsed tables, gate verdicts, and D201 evidence predicates for profiles c1/c5/c10 and payloads 250/450/520/900
> - Modifies `executeBenchmark` to embed a `ci_gate` block in the receipt and set a non-zero exit code whenever the gate produces any failure
> - Behavioral Change: `executeBenchmark` now exits non-zero on MCP CLI fallback or unattested transports, not only on fatal errors or `error_count > 0`; consumers reading the receipt must handle the new `ci_gate` field
>
> 
>
> <sup><a href="https://app.macroscope.com">Macroscope</a> summarized 1d75855.</sup>
> 
>

## Summary by CodeRabbit

* **New Features**
  * Added benchmark validation for CLI fallbacks, unsupported transport evidence, and other gating failures.
  * Added publication output for the Phase 1 BEFORE benchmark matrix, including measurement tags and D201 evidence.
  * Benchmark receipts now include a clear CI gate verdict.

* **Bug Fixes**
  * Benchmark runs now exit with a failure status whenever any configured gate fails, not only on fatal errors.

* **Tests**
  * Added coverage for gate detection, publication formatting, collapsed results, evidence details, and inconclusive outcomes.

### CODE REVIEW

**Comment by EtanHey** [comment]
## Run 10 Phase 1 BEFORE: RED

Source: 53df362d22a02f85ff948cdb5b6b70d1e8c8d96a at 2026-08-31T12:25:31.400Z. Surface-mode transport is **UNTRUSTED** under D180; reported socket provenance is not treated as attested transport.

| operation | profile | payload | client | measurement | status | reason | attempts | successes | failure % | p50 ms | p95 ms | transport |
|---|---:|---:|---|---|---|---|---:|---:|---:|---:|---:|---|
| send_to | c1 | 250 | cli | single_shot | NOT_COMPARABLE | FAIRNESS_CONTRACT | — | — | — | — | — | none |
| send_to | c1 | 450 | cli | single_shot | NOT_COMPARABLE | FAIRNESS_CONTRACT | — | — | — | — | — | none |
| send_to | c1 | 520 | mcp | sampled | FAIL | CLI_FALLBACK | 12 | 12 | 0 | 531.37 | 1572.62 | UNTRUSTED (D180) |
| send_to | c1 | 520 | cli | single_shot | NOT_COMPARABLE | FAIRNESS_CONTRACT | — | — | — | — | — | none |
| send_to | c1 | 900 | mcp | sampled | FAIL | CLI_FALLBACK | 12 | 12 | 0 | 484.9 | 1343.14 | UNTRUSTED (D180) |
| send_to | c1 | 900 | cli | single_shot | NOT_COMPARABLE | FAIRNESS_CONTRACT | — | — | — | — | — | none |
| send_to | c5 | 250 | cli | single_shot | NOT_COMPARABLE | FAIRNESS_CONTRACT | — | — | — | — | — | none |
| send_to | c5 | 450 | cli | single_shot | NOT_COMPARABLE | FAIRNESS_CONTRACT | — | — | — | — | — | none |
| send_to | c5 | 520 | mcp | sampled | FAIL | CLI_FALLBACK, OPERATION_ERROR | 60 | 45 | 25 | 819.45 | 1069.13 | UNTRUSTED (D180) |
| send_to | c5 | 520 | cli | single_shot | NOT_COMPARABLE | FAIRNESS_CONTRACT | — | — | — | — | — | none |
| send_to | c5 | 900 | mcp | sampled | FAIL | CLI_FALLBACK, OPERATION_ERROR | 60 | 46 | 23.33 | 1234.46 | 1600.52 | UNTRUSTED (D180) |
| send_to | c5 | 900 | cli | single_shot | NOT_COMPARABLE | FAIRNESS_CONTRACT | — | — | — | — | — | none |
| send_to | c10 | 250 | cli | single_shot | NOT_COMPARABLE | FAIRNESS_CONTRACT | — | — | — | — | — | none |
| send_to | c10 | 450 | cli | single_shot | NOT_COMPARABLE | FAIRNESS_CONTRACT | — | — | — | — | — | none |
| send_to | c10 | 520 | mcp | sampled | FAIL | CLI_FALLBACK, OPERATION_ERROR | 120 | 110 | 8.33 | 1753.81 | 2053.75 | UNTRUSTED (D180) |
| send_to | c10 | 520 | cli | single_shot | NOT_COMPARABLE | FAIRNESS_CONTRACT | — | — | — | — | — | none |
| send_to | c10 | 900 | mcp | sampled | FAIL | CLI_FALLBACK, OPERATION_ERROR | 120 | 112 | 6.67 | 2198.73 | 2977.9 | UNTRUSTED (D180) |
| send_to | c10 | 900 | cli | single_shot | NOT_COMPARABLE | FAIRNESS_CONTRACT | — | — | — | — | — | none |
| list_surfaces | c5 | - | cli | sampled | FAIL | OPERATION_ERROR | 60 | 57 | 5 | 278.26 | 416.81 | cli:57,unknown:3 |
| list_surfaces | c10 | - | cli | sampled | FAIL | OPERATION_ERROR | 120 | 106 | 11.67 | 655.48 | 859.84 | cli:106,unknown:14 |

16 rows unchanged.

<details>
<summary>Full 36-row BEFORE table</summary>

| operation | profile | payload | client | measurement | status | reason | attempts | successes | failure % | p50 ms | p95 ms | transport |
|---|---:|---:|---|---|---|---|---:|---:|---:|---:|---:|---|
| send_to | c1 | 250 | mcp | sampled | PASS | — | 12 | 12 | 0 | 204.84 | 225.63 | UNTRUSTED (D180) |
| send_to | c1 | 250 | cli | single_shot | NOT_COMPARABLE | FAIRNESS_CONTRACT | — | — | — | — | — | none |
| send_to | c1 | 450 | mcp | sampled | PASS | — | 12 | 12 | 0 | 299.74 | 503.5 | UNTRUSTED (D180) |
| send_to | c1 | 450 | cli | single_shot | NOT_COMPARABLE | FAIRNESS_CONTRACT | — | — | — | — | — | none |
| send_to | c1 | 520 | mcp | sampled | FAIL | CLI_FALLBACK | 12 | 12 | 0 | 531.37 | 1572.62 | UNTRUSTED (D180) |
| send_to | c1 | 520 | cli | single_shot | NOT_COMPARABLE | FAIRNESS_CONTRACT | — | — | — | — | — | none |
| send_to | c1 | 900 | mcp | sampled | FAIL | CLI_FALLBACK | 12 | 12 | 0 | 484.9 | 1343.14 | UNTRUSTED (D180) |
| send_to | c1 | 900 | cli | single_shot | NOT_COMPARABLE | FAIRNESS_CONTRACT | — | — | — | — | — | none |
| send_to | c5 | 250 | mcp | sampled | PASS | — | 60 | 60 | 0 | 926 | 1050.78 | UNTRUSTED (D180) |
| send_to | c5 | 250 | cli | single_shot | NOT_COMPARABLE | FAIRNESS_CONTRACT | — | — | — | — | — | none |
| send_to | c5 | 450 | mcp | sampled | PASS | — | 60 | 60 | 0 | 900.85 | 1062.74 | UNTRUSTED (D180) |
| send_to | c5 | 450 | cli | single_shot | NOT_COMPARABLE | FAIRNESS_CONTRACT | — | — | — | — | — | none |
| send_to | c5 | 520 | mcp | sampled | FAIL | CLI_FALLBACK, OPERATION_ERROR | 60 | 45 | 25 | 819.45 | 1069.13 | UNTRUSTED (D180) |
| send_to | c5 | 520 | cli | single_shot | NOT_COMPARABLE | FAIRNESS_CONTRACT | — | — | — | — | — | none |
| send_to | c5 | 900 | mcp | sampled | FAIL | CLI_FALLBACK, OPERATION_ERROR | 60 | 46 | 23.33 | 1234.46 | 1600.52 | UNTRUSTED (D180) |
| send_to | c5 | 900 | cli | single_shot | NOT_COMPARABLE | FAIRNESS_CONTRACT | — | — | — | — | — | none |
| send_to | c10 | 250 | mcp | sampled | PASS | — | 120 | 120 | 0 | 1971.2 | 3289.22 | UNTRUSTED (D180) |
| send_to | c10 | 250 | cli | single_shot | NOT_COMPARABLE | FAIRNESS_CONTRACT | — | — | — | — | — | none |
| send_to | c10 | 450 | mcp | sampled | PASS | — | 120 | 120 | 0 | 1712.22 | 2557.62 | UNTRUSTED (D180) |
| send_to | c10 | 450 | cli | single_shot | NOT_COMPARABLE | FAIRNESS_CONTRACT | — | — | — | — | — | none |
| send_to | c10 | 520 | mcp | sampled | FAIL | CLI_FALLBACK, OPERATION_ERROR | 120 | 110 | 8.33 | 1753.81 | 2053.75 | UNTRUSTED (D180) |
| send_to | c10 | 520 | cli | single_shot | NOT_COMPARABLE | FAIRNESS_CONTRACT | — | — | — | — | — | none |
| send_to | c10 | 900 | mcp | sampled | FAIL | CLI_FALLBACK, OPERATION_ERROR | 120 | 112 | 6.67 | 2198.73 | 2977.9 | UNTRUSTED (D180) |
| send_to | c10 | 900 | cli | single_shot | NOT_COMPARABLE | FAIRNESS_CONTRACT | — | — | — | — | — | none |
| read_screen | c1 | - | mcp | sampled | PASS | — | 12 | 12 | 0 | 40.1 | 50.42 | socket:12 |
| read_screen | c1 | - | cli | sampled | PASS | — | 12 | 12 | 0 | 677.3 | 730.73 | cli:12 |
| read_screen | c5 | - | mcp | sampled | PASS | — | 60 | 60 | 0 | 210.82 | 268.58 | socket:60 |
| read_screen | c5 | - | cli | sampled | PASS | — | 60 | 60 | 0 | 742.63 | 841.04 | cli:60 |
| read_screen | c10 | - | mcp | sampled | PASS | — | 120 | 120 | 0 | 339.14 | 370.67 | socket:120 |
| read_screen | c10 | - | cli | sampled | PASS | — | 120 | 120 | 0 | 735.83 | 842.51 | cli:120 |
| list_surfaces | c1 | - | mcp | sampled | PASS | — | 12 | 12 | 0 | 15.45 | 30.19 | socket:12 |
| list_surfaces | c1 | - | cli | sampled | PASS | — | 12 | 12 | 0 | 257.75 | 292.45 | cli:12 |
| list_surfaces | c5 | - | mcp | sampled | PASS | — | 60 | 60 | 0 | 24.41 | 25.32 | socket:60 |
| list_surfaces | c5 | - | cli | sampled | FAIL | OPERATION_ERROR | 60 | 57 | 5 | 278.26 | 416.81 | cli:57,unknown:3 |
| list_surfaces | c10 | - | mcp | sampled | PASS | — | 120 | 120 | 0 | 273.46 | 816.89 | socket:120 |
| list_surfaces | c10 | - | cli | sampled | FAIL | OPERATION_ERROR | 120 | 106 | 11.67 | 655.48 | 859.84 | cli:106,unknown:14 |

</details>

## D201 mechanism evidence

- Boundary: 250/450 = 0% at c1, c5, and c10 (c1 0%/0%; c5 0%/0%; c10 0%/0%).
- Concurrency: c1 = 0% at both 520 and 900 characters (0%/0%).
- Non-monotonic curve: c5 520/900 = 25%/23.33% at p50 819.45/1234.46 ms; c10 520/900 = 8.33%/6.67% at p50 1753.81/2198.73 ms. c10 fails less than c5 while latency rises substantially.
- Interpretation: this is evidence against simple capacity exhaustion and for a bounded race window. It narrows candidates; it does not identify the racing party. Read the buffer store capacity and eviction policy before proposing a paste-path fix.

— cmuxlayerCodex-2c968dd2 (worker) · codex/gpt-5.6-sol

**Comment by EtanHey** [comment]
@coderabbitai review
@codex review
@deepsourcebot review
@greptileai review

Please review exact head `196162d38394f57cd3fb557908abe22a86974457` for false-green fallback gating, publication accuracy, and D180/D201 semantics.

— cmuxlayerCodex-2c968dd2 (worker) · codex/gpt-5.6-sol

File: scripts/bench-e2e.mjs [hunk 1]
```diff
@@ -1918,6 +1918,208 @@ function compactCounts(counts) {
...
+    if (row.client !== "mcp") continue;
+    const evidence = [];
+    const unattested = [];
+    if (row.inferred_transport === "cli") {
+      evidence.push("inferred_transport=cli");
+    }
+    let hasAttestedTransport = ["cli", "socket"].includes(
+      row.inferred_transport,
+    );
+    for (const field of ["transport_counts", "reported_transport_counts"]) {
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply scripts/bench-e2e.mjs]: Resolved in `5bb52eb`: only raw `transport_counts.socket|cli` establishes attestation; inferred and reported transports remain diagnostic/fallback evidence. Explicit D180-untrusted surface rows retain a bounded exemption keyed by `transport_trust: untrusted` plus `UNTRUSTED_D180`. The regression now covers an MCP row that has only inferred/reported socket evidence and requires `unattested transport` RED. Focused harness: 113/113; protected hook: 155 files / 3,696 passed / 1 skipped.

— cmuxlayerCodex-2c968dd2 (worker) · codex/gpt-5.6-sol

File: scripts/bench-e2e.mjs [hunk 2]
```diff
@@ -1918,6 +1918,208 @@ function compactCounts(counts) {
...
+  const visibleRows = rows.filter(
+    (row) => publicationStatus(row) !== "PASS" || row.moved === true,
+  );
+  const unchangedCount = rows.length - visibleRows.length;
+  const gateFailures = benchmarkGateFailures(rows, receipt.fatal_error);
+  return [
+    "<!-- cmuxlayer-run10-phase1-before -->",
+    `## Run 10 Phase 1 BEFORE: ${gateFailures.length === 0 ? "GREEN" : "RED"}`,
+    "",
+    `Source: ${receipt.git_head ?? "unknown"} at ${receipt.started_at ?? "unknown"}. Surface-mode transport is **UNTRUSTED** under D180; reported socket provenance is not treated as attested transport.`,
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply scripts/bench-e2e.mjs:2141]: Resolved in `e250158`: a non-null `fatal_error` is normalized and rendered as `Fatal gate failure: …` in the visible section above the table. The regression first reproduced the hidden RED cause, then passed with the repair. Focused harness: 113/113; both typechecks and D138 green; protected hook: 155 files / 3,696 passed / 1 skipped.

— cmuxlayerCodex-2c968dd2 (worker) · codex/gpt-5.6-sol

File: scripts/bench-e2e.mjs [hunk 3]
```diff
@@ -1918,6 +1918,220 @@ function compactCounts(counts) {
...
+    `${unchangedCount} rows unchanged.`,
+    "",
+    "<details>",
+    `<summary>Full ${rows.length}-row BEFORE table</summary>`,
+    "",
+    renderPublicationRows(rows),
+    "",
+    "</details>",
+    "",
+    d201Evidence(rows),
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply scripts/bench-e2e.mjs:2155]: Fixed at f340903c94969eaa7d4d817d750fa750d0a9accf. Partial fatal receipts now render a D201 INCONCLUSIVE section listing every missing required sampled row instead of throwing; complete 12-row D201 evaluation is unchanged. Regression: bun test tests/bench-e2e.test.ts => 113 pass, 0 fail. Protected push hook: 155 files passed, 3696 tests passed, 1 skipped, exit 0.\n\n— cmuxlayerCodex-2c968dd2 (worker) · codex/gpt-5.6-sol\n

File: scripts/bench-e2e.mjs [hunk 4]
```diff
@@ -1918,6 +1918,243 @@ function compactCounts(counts) {
...
+      row("c10", payload).failure_rate_pct <
+      row("c5", payload).failure_rate_pct,
+  );
+  const c10LatencyRises = [520, 900].every(
+    (payload) => row("c10", payload).p50_ms > row("c5", payload).p50_ms,
+  );
+  const observed = [
+    "## D201 mechanism evidence",
+    "",
+    `- Boundary: 250/450 = 0% at c1, c5, and c10 (${profiles.map((profile) => `${profile} ${row(profile, 250).failure_rate_pct}%/${row(profile, 450).failure_rate_pct}%`).join("; ")}).`,
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply scripts/bench-e2e.mjs]: Fixed at 1d758552686944a86fa2d749187b220096add0a1. The boundary label is now affirmative only when boundaryZero holds; otherwise it reports neutral observed rates. The same truth condition now covers c1 and the c5/c10 rate-latency conclusion. Accepted-receipt output remains byte-identical. Regression: 113 pass, 0 fail; protected hook: 155 files, 3696 passed, 1 skipped, exit 0.\n\n— cmuxlayerCodex-2c968dd2 (worker) · codex/gpt-5.6-sol\n

File: scripts/bench-e2e.mjs [hunk 5]
```diff
@@ -1918,6 +1918,243 @@ function compactCounts(counts) {
...
+    `- Concurrency: c1 = 0% at both 520 and 900 characters (${row("c1", 520).failure_rate_pct}%/${row("c1", 900).failure_rate_pct}%).`,
+    `- Non-monotonic curve: c5 520/900 = ${row("c5", 520).failure_rate_pct}%/${row("c5", 900).failure_rate_pct}% at p50 ${row("c5", 520).p50_ms}/${row("c5", 900).p50_ms} ms; c10 520/900 = ${row("c10", 520).failure_rate_pct}%/${row("c10", 900).failure_rate_pct}% at p50 ${row("c10", 520).p50_ms}/${row("c10", 900).p50_ms} ms. c10 fails less than c5 while latency rises substantially.`,
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply scripts/bench-e2e.mjs]: Fixed at 1d758552686944a86fa2d749187b220096add0a1. Each D201 summary is now conditional on its predicate: failed boundary and c1 conditions use neutral observed-rate labels, and the c10-below-c5/latency conclusion is emitted only when both predicates hold. A combined counterexample regression proves none of the three false claims is rendered. Accepted-receipt output remains byte-identical. Harness: 113 pass; protected hook: 155 files, 3696 passed, 1 skipped, exit 0.\n\n— cmuxlayerCodex-2c968dd2 (worker) · codex/gpt-5.6-sol\n
